daemons/fenced/pacemaker-fenced.c

```diff
@@ -35,7 +35,7 @@
 
 #include <crm/cib/internal.h>
 #include <crm/pengine/status.h>
-#include <pacemaker-internal.h>
+#include <pacemaker.h>
```

Comment:
Now that libpacemaker will have both private and public APIs, we'll need to reserve pacemaker.h for the public API.

Taking a fresh look at naming in light of recent discussions, it looks like we're moving toward an eventual library split of:
* libpacemaker: high-level (tool) functionality (likely including the current libpe_status and libpe_rules)
* libpacemaker-cluster: the current libcrmcluster (which depends on corosync libraries)
* (maybe) libpacemaker-curses: crm_mon's curses functionality (which depends on curses libraries)
* everything else either in one "low-level" library, or split into libpacemaker-common (lowest level, i.e. most of libcrmcommon and libcrmservice), libpacemaker-client (client APIs, i.e. libcib, libstonithd, and liblrmd, and some of libcrmcommon), and libpacemaker-server (the few functions that are only needed by the daemons).

Looking at that, I'm thinking we should abandon the idea of using a unique prefix for each library, and just use pcmk for everything. It's especially difficult since up to four libraries would start with a "c". :-) I see other projects with multiple libraries that don't tie function prefixes to libraries.

It's also questionable whether we should keep using the include/crm subdirectory. Distros tend to install to /usr/include/pacemaker, so currently everything is in /usr/include/pacemaker/crm/*.h. It's an unnecessary level, and would look even weirder if we started calling it "pacemaker" e.g. /usr/include/pacemaker/pacemaker/pacemaker.h! Looking at other common projects, most don't worry about adding another level, so I think we can abandon crm for the new libraries.

I'm open to suggestions for how to name the internal headers. Currently we tend to put "_internal" in the name somewhere (though not consistently). We could continue with that, or keep the pcmki prefix just for this purpose, or extend the double underbar convention to header file names, or something else.

Bottom line, maybe this header should stay named pacemaker-internal.h, or pacemaker_internal.h.

Reply:
> (maybe) libpacemaker-curses: crm_mon's curses functionality (which
> depends on curses libraries)

This is symptomatic to what I find a bad functional decomposition of
the problem.  More ideally, front-ends would be rather separa(ted,ble}
formatters of a somehow unified data model, see also
https://bugs.clusterlabs.org/show_bug.cgi?id=5376#c3

Especially formatting to the terminal is rather a unique one-off
functionality presumably having just a single client, hence belongs
to such program itself (or a private library if need be, waiting
for other external client wanting to yield 1:1 data-to-terminal
output, which doesn't sound very probable).

Reply:
Exactly, #1707 is indeed a unified output formatting model with each output type having its own formatter. Of course, curses output requires linking against curses libraries, so we can't put that portion of the code into a common library.

The end goal for the tools, including crm_mon, is to be able to call their functionality as C API. crm_mon's curses capability makes it a special case. We want to have all of crm_mon's capabilities except curses output as an API in libpacemaker, but have crm_mon itself able to register the curses formatter before calling that same API. Whether the curses formatter is in the crm_mon code, a library private to crm_mon, or an internal library that could be used by future tools is an implementation detail that we can address at that time.

Reply:
Ah, was worried that in-program (most appropriate here, I think,
shared library is for reusing, single user = no reusing) hosting
was on the way out along the actual reusable parts.

Btw. maybe would be interesting to put related clusters (sic) of
private-use-only symbols behind lightly used ones and, at the same
time, without any logging (since it would then defeat "singleton"
principle of libqb logging arrangement at least with linker section
involved, I think, impacting efficiency) into a (single) static library
on its own to be linked with _all_ the binary artefacts (will become
a no-op when they don't ask for any such symbol).  This way, we would
keep the dynamic symbol resolution clutter and overhead (also durable
burden with files in locations where publiclz reusable libraries are
expected) minimal. The static size overhead won't be noticable and
pacemaker parts are mostly required to stay in sync anyway.
Code locality would increase (as opposed to private shared libraries),
which is good.  Afterall that's the path modern languages take
(Go, Rust).

Using private shared libraries "because we can and know to do it,
since we have also actual public libraries" is not a very profound
justification to do the same for the private purposes.
Of course, there's also this very basic modules reusing approach,
there's nothing bad about that either, but can be administratively
more clumsy (price paid for a full control).

Reply:
libpacemaker isn't solely about reusability; a major goal is to expose the tool functionality as C API so that external projects can use them to create language bindings, API servers, high-level tools, etc.

Reply:
I didn't mean to put `libpacemaker` and static library into one equation,
it was meant to address some parts of what would remain "private yet still
shared libs" (such as ncurses-based output formatters).